lib/default.nix

```diff
@@ -393,4 +393,96 @@ in {
   # one in the list will be used.
   combineFlakes = sep: prefixAndFlakes: builtins.foldl' addFlakes {}
     (lib.mapAttrsToList (prefix: flake: prefixFlake prefix sep flake) prefixAndFlakes);
+
+  mkFlake = project: packages: coverageProjectModule:
+    let
+      packageNames = project: builtins.attrNames (project.args.flake.packages project.hsPkgs);
+      checkedProject = project.appendModule { checkMaterialization = true; };
+      coverageProject = project.appendModule [
+        coverageProjectModule
+        {
+          modules = [{
+            packages = lib.genAttrs (packageNames project)
+              (_: { doCoverage = lib.mkDefault true; });
+          }];
+        }
+      ];
+      checks = builtins.listToAttrs (
+        lib.concatMap (packageName:
+          let package = project.hsPkgs.${packageName};
+          in lib.mapAttrsToList (n: v:
+              { name = "${packageName}:test:${n}"; value = v; })
+            (lib.filterAttrs (_: v: lib.isDerivation v) (package.checks))
+        ) (packageNames project));
+      ciJobs = {
+          # Run all the tests
+          inherit checks;
+          # Build tools and cache tools needed for the project
+          inherit (project) roots;
+          # Also build and cache any tools in the `devShell`
+          devShell = project.shell;
+          # Run HPC on the tests
+          coverage = builtins.listToAttrs (lib.concatMap (packageName: [{
+            name = packageName;
+            value = coverageProject.hsPkgs.${packageName}.coverageReport;
+          }]) (packageNames coverageProject));
+        }
+        # Build the plan-nix and check it if materialized
+        // lib.optionalAttrs (checkedProject ? plan-nix) {
+          plan-nix = checkedProject.plan-nix;
+        }
+        # Build the stack-nix and check it if materialized
+        // lib.optionalAttrs (checkedProject ? stack-nix) {
+          stack-nix = checkedProject.stack-nix;
+        };
+    in {
+      # Used by:
+      #   `nix build .#pkg-name:lib:pkg-name`
+      #   `nix build .#pkg-name:lib:sublib-name`
+      #   `nix build .#pkg-name:exe:exe-name`
+      #   `nix build .#pkg-name:test:test-name`
+      packages = builtins.listToAttrs (
+        lib.concatMap (packageName:
+          let package = project.hsPkgs.${packageName};
+          in lib.optional (package.components ? library)
```

Comment:
Again, I'm kind of surprised we're not using the `collect*` functions? Is it that they don't give the attribute keys that we want? Maybe we should change those functions, then?